torchtune/data/_utils.py

```diff
@@ -142,3 +150,171 @@ def format_content_with_images(
+            num_replicas=world_size,
+            rank=rank,
+            shuffle=shuffle,
+            seed=seed,
+        )
+        node = SamplerWrapper(sampler)
+        transform = chain(dataset.__getitem__, transform)  # type: ignore
+
+    node = ParallelMapper(
+        node, map_fn=transform, num_workers=num_workers, method=parallel_method
+    )
+
+    return node
+
+
+@requires_torchdata
+def get_multi_dataset(
+    datasets: dict[str, DatasetType],
+    weights: dict[str, float],
+    stop_criteria: str = "CYCLE_UNTIL_ALL_DATASETS_EXHAUSTED",
+    seed: int = 0,
+) -> DatasetType:
+    """
+    Given a dictionary of datasets and their corresponding weights, return a dataset that
+    samples from the given datasets according to the specified weights.
+
+    Args:
+        datasets (Dict[str, Any]): dictionary of datasets
+        weights (Optional[Dict[str, float]]): dictionary of weights for each dataset. If not
+        stop_criteria (str): stop criteria for the sampler. Default "CYCLE_UNTIL_ALL_DATASETS_EXHASTED".
+            see also: torchdata.nodes.StopCriteria
+        seed: (int): seed for the random number generator. Default 0.
+
+    """
+    from torchdata.nodes import MultiNodeWeightedSampler
+
+    return MultiNodeWeightedSampler(
+        source_nodes=datasets,
+        weights=weights,
+        stop_criteria=stop_criteria,
+        seed=seed,
+    )
+
+
+@requires_torchdata
+def get_dataloader(
+    dataset: DatasetType,
+    model_transform: Transform,
+    batch_size: int,
+    collate_fn: Optional[Callable[[Any], Any]] = None,
+    packed: bool = False,
+    drop_last: bool = True,
+    num_workers: int = 0,
+    parallel_method: Literal["process", "thread"] = "thread",
+    prefetch_factor: Optional[int] = 4,
+    pin_memory: bool = False,
+) -> Loader:
+    """
+    This will configure TorchData Nodes to approximate torch.utils.data.DataLoader.
+    Given a dataset, apply model_transform (eg tokenization), batching, collation,
+    memory pinning, and pre-fetching.
+
+    Args:
+        dataset (DatasetType): dataset to load. May be a MultiNodeWeightedSampler
+        model_transform (Transform): model transform to apply to the samples of the dataset
+        batch_size (int): batch size
+        collate_fn (Callable[[Any], Any]): collate function to apply to the samples of the dataset
+        packed (bool): whether to pack the dataset. Default is False. Not supported yet.
+        drop_last (bool): whether to drop the last batch. Default is True.
+        num_workers (int): number of workers to use for loading the dataset. Default is 0 (no parallelism
+        parallel_method (Literal["process", "thread"]): method to use for parallelism. Default is "thread".
+        prefetch_factor (Optional[int]): number of batches to prefetch. Default is 4.
+        pin_memory (bool): whether to pin memory. Default is False.
+    """
+    if packed:
+        raise ValueError("Multimodal datasets don't support packing yet.")
```

Comment:
Ultimately this isn't the right place for this check, right? I assume we will need to do it somewhere else (or at least in another way) once we onboard text datasets

Reply:
I think you're right, I'm assuming that packer behaviour is going to need to be configurable, so we'll probably need to change this signature entirely anyways. I'll just drop it as an option and catch this in lora_finetune_distributed_multi_dataset for now, wdyt? @ebsmothers

Reply:
Yeah that works. We can revisit when we're migrating text recipes